### PR TITLE
Set the cloud token in service discovery credentials when none exist for the cloud hostname

### LINKED ISSUE #32665: `unauthorized` error when rendering a plan with cloud integration token attribute
### Terraform Version

```shell
Terraform v1.4.0-beta1
on darwin_amd64
```

### Terraform Configuration Files

```terraform
terraform {
  cloud {
    organization = "bcroft"
    token = "blah"

    workspaces {
      name = "example-null"
    }
  }
}

resource "null_resource" "null" {
  triggers = {
    "key" = "value2"
  }
}
```

### Debug Output

```
$ TF_LOG=debug terraform apply
2023-02-10T13:25:58.100-0700 [INFO]  Terraform version: 1.4.0 beta1
2023-02-10T13:25:58.101-0700 [DEBUG] using github.com/hashicorp/go-tfe v1.18.0
2023-02-10T13:25:58.101-0700 [DEBUG] using github.com/hashicorp/hcl/v2 v2.16.0
2023-02-10T13:25:58.101-0700 [DEBUG] using github.com/hashicorp/terraform-config-inspect v0.0.0-20210209133302-4fd17a0faac2
2023-02-10T13:25:58.101-0700 [DEBUG] using github.com/hashicorp/terraform-svchost v0.1.0
2023-02-10T13:25:58.101-0700 [DEBUG] using github.com/zclconf/go-cty v1.12.1
2023-02-10T13:25:58.101-0700 [INFO]  Go runtime version: go1.19.4
2023-02-10T13:25:58.101-0700 [INFO]  CLI args: []string{"/usr/local/Cellar/tfenv/3.0.0/versions/1.4.0-beta1/terraform", "apply"}
2023-02-10T13:25:58.101-0700 [DEBUG] Attempting to open CLI config file: /Users/brandonc/hashicorp/terraform-provider-tfe/provider_tfe_override.tfrc
2023-02-10T13:25:58.101-0700 [INFO]  Loading CLI configuration from /Users/brandonc/hashicorp/terraform-provider-tfe/provider_tfe_override.tfrc
2023-02-10T13:25:58.102-0700 [DEBUG] Not reading CLI config directory because config location is overridden by environment variable
2023-02-10T13:25:58.103-0700 [DEBUG] checking for credentials in "/Users/brandonc/.terraform.d/plugins"
2023-02-10T13:25:58.103-0700 [DEBUG] Explicit provider installation configuration is set
2023-02-10T13:25:58.103-0700 [INFO]  CLI command args: []string{"apply"}
2023-02-10T13:25:58.107-0700 [DEBUG] Service discovery for app.terraform.io at https://app.terraform.io/.well-known/terraform.json
2023-02-10T13:25:58.345-0700 [DEBUG] Service discovery for app.terraform.io aliased as localterraform.com
2023-02-10T13:25:58.669-0700 [DEBUG] Provider registry.terraform.io/hashicorp/tfe is overridden by dev_overrides
2023-02-10T13:25:58.714-0700 [DEBUG] Provider registry.terraform.io/hashicorp/tfe is overridden to load from /Users/brandonc/hashicorp/terraform-provider-tfe
2023-02-10T13:25:58.714-0700 [DEBUG] checking for provisioner in "."
2023-02-10T13:25:58.715-0700 [DEBUG] checking for provisioner in "/usr/local/Cellar/tfenv/3.0.0/versions/1.4.0-beta1"
2023-02-10T13:25:58.715-0700 [DEBUG] checking for provisioner in "/Users/brandonc/.terraform.d/plugins"
2023-02-10T13:25:58.716-0700 [DEBUG] Provider registry.terraform.io/hashicorp/tfe is overridden by dev_overrides
2023-02-10T13:25:58.861-0700 [INFO]  cloud: starting Apply operation
Running apply in Terraform Cloud. Output will stream here. Pressing Ctrl-C
will cancel the remote apply if it's still pending. If the apply started it
will stop streaming the logs, but will not stop the apply running remotely.

Preparing the remote apply...

To view this run in a browser, visit:
https://app.terraform.io/app/bcroft/example-null/runs/run-4mVmAcRkdssRyAYk

Waiting for the plan to start...

Terraform v1.4.0-beta1
on linux_amd64
Initializing plugins and modules...
2023-02-10T13:26:49.484-0700 [DEBUG] performing request: method=GET url=https://app.terraform.io/api/v2/plans/plan-gVkqwkgcAWQZdxUL/json-output-redacted
╷
│ Error: unauthorized
│
│
╵
```

### Expected Behavior

The plan diff should render.

### Actual Behavior

An `unauthorized` error occurred.

### Steps to Reproduce

1. Ensure that no credentials are supplied in the CLI config
2. Ensure token is set in cloud block instead
3. `terraform init`
4. `terraform apply`

### Additional Context

The token set in config is not being utilized [by the renderPlanLogs function](https://github.com/hashicorp/terraform/blob/main/internal/cloud/backend_plan.go#L435) and instead only the CLI config credentials token is being used to configure the request to the redacted json endpoint. It would be a good idea to configure the token on the backend instead of re-using the token() function, which doesn't consult the cloud block.

### References

_No response_

### PR DESCRIPTION
**Pull request by brandonc**
Before, it was not possible to access services credentials given only the cloud block (not CLI config). Things like private registry module sources in Terraform Cloud must use `terraform login` to successfully authenticate.

This change shims the credentials source with the cloud token if no creds exist for the cloud hostname. This keeps the existing behavior of preferring terraform login for module resolution, but also is a convenience to callers of backend.token(), private registry module developers, and future services provided by Terraform Cloud.

Consider this example config:

```terraform
terraform {
  cloud {
    organization = "hashicorp"
    token = "MY_TOKEN"
    workspaces {
      name = "cli-onefourtesting"
    }
  }
}

module "my-aws-module" {
  source  = "app.terraform.io/hashicorp/simple-s3-cdn/aws"
}
```

Even though the cloud token is given in the cloud block, authorization fails when initializing the module found in the cloud private registry, because only the cloud backend package is configured with the token, not the underlying service discovery.

This also fixes #32665 because subsequent calls to backend token() include the cloud block token creds.

## Target Release

1.4.0

### ENHANCEMENTS

Cloud block token credentials are now used for private registry authentication when using the same cloud hostname

### CODE REVIEW

File: internal/cloud/backend.go [hunk 1]
```diff
@@ -245,6 +246,20 @@ func (b *Cloud) Configure(obj cty.Value) tfdiags.Diagnostics {
 	var token string
 	if val := obj.GetAttr("token"); !val.IsNull() {
 		token = val.AsString()
+
+		if token != "" {
+			cliToken, err := b.token()
+			// Service discovery is the canonical source for credentials, but in the absence of
```
Comment (annawinkler) [inline internal/cloud/backend.go:252]: 💯 Great comment!

File: internal/cloud/backend.go [hunk 2]
```diff
@@ -245,6 +246,20 @@ func (b *Cloud) Configure(obj cty.Value) tfdiags.Diagnostics {
...
+			// credentials for the cloud hostname in that source, make an effort to use the token set by
+			// cloud config. This allows cloud credentials to be used for things like private
+			// registry module sources authentication.
+			if cliToken == "" && err == nil {
+				log.Printf("[DEBUG] No CLI credentials found for %q, shimming cloud token", b.hostname)
+				if err := b.services.CredentialsSource().StoreForHost(svchost.Hostname(b.hostname), auth.HostCredentialsToken(token)); err != nil {
+					log.Printf("[DEBUG] Failed to services credentials source with cloud token: %s", err)
```
Comment (brandonc) [inline internal/cloud/backend.go:259]: I accidentally a word.

**Review by sebasslash — APPROVED** [review]
Awesome. I'll close #32666 once merged.

**Comment by nfagerlund** [comment]
I put this through its paces, and it mostly works as hoped-for. However, I found a weird unintended behavior, and @brandonc didn't feel good about rushing it in under those circumstances, so he's going to re-open #32666 and circle back around in a later release for an improved version of this holistic fix. 

The issue: If the `~/.terraform.d/credentials.tfrc.json` login file **lacks** a token for the host but the `cloud` block in the config specifies one directly, Terraform will auto-login — writing the config's token into the login credentials file. It does this even if the token is bogus, which then requires a `terraform logout` for the host to fix. 

By contrast, if you specify a `TF_TOKEN_app_terraform_io` env var with no login credentials, it'll use the token from the env var without adding it to the login file. That seems to be what we'd want to have happen with a token in the config... and hopefully we'll be able to imitate the implementation of that for this.

Summary of the testing I did: 

Login creds | TF_TOKEN_env_var | Cloud block | Init results
-|-|-|-
none | none | good | Works, but writes new login credentials
good | none | none | Works
none | good | none | Works
good | none | good | Works, doesn't overwrite existing good login token
bad | none | good | Fails module and provider install (bad token wins)
good | none | bad | Installs module, fails before provider install (good token used some places, but not others)
none | good | bad | Same as previous
none | none | bad | Fails mod/provider install, writes bad token to login file

In a few cases (only cloud block, only login file), I also tested with localterraform.com in module/providers sources instead of app.terraform.io. No behavior differences observed, so that's cool.

**Comment by nfagerlund** [comment]
For what it's worth, I consider that unintended behavior to be really mild, and if we find that it's not practical to get an equivalent of what the env var does, then I think it ought to be fine to ship with that rough edge. But it's probably worth making an attempt at avoiding it first.

**Pull request closed without merging**